consensus/core/src/leader_scoring_strategy.rs

```diff
@@ -0,0 +1,449 @@
+            if subdag.is_vote(&potential_vote, leader_block) {
+                let authority = potential_vote.author();
+                tracing::info!(
+                    "Found a vote {} for leader {leader_block} from authority {authority}",
+                    potential_vote.reference()
+                );
+                tracing::info!(
+                    "[{}] scores +1 reputation for {authority}!",
+                    subdag.context.own_index
+                );
+                scores_per_authority[authority] += 1;
+            }
+        }
+
+        scores_per_authority
+    }
+
+    fn leader_scoring_round_range(&self, min_round: u32, max_round: u32) -> Range<u32> {
+        assert!(min_round < max_round);
+        min_round..max_round
+    }
+}
+
+// This scoring strategy will give one point to any certificates for the leader.
+pub(crate) struct CertificateScoringStrategy {}
+
+impl ScoringStrategy for CertificateScoringStrategy {
+    fn calculate_scores_for_leader(
+        &self,
+        subdag: &UnscoredSubdag,
+        leader_slot: Slot,
+        committer: &BaseCommitter,
+    ) -> Vec<u64> {
+        let num_authorities = subdag.context.committee.size();
+        let mut scores_per_authority = vec![0_u64; num_authorities];
+
+        let wave = committer.wave_number(leader_slot.round);
+        let decision_round = committer.decision_round(wave);
+
+        let leader_blocks = subdag.get_blocks_at_slot(leader_slot);
+
+        if leader_blocks.is_empty() {
+            tracing::info!("[{}] No block for leader slot {leader_slot} in this set of unscored committed subdags, skip scoring", subdag.context.own_index);
+            return scores_per_authority;
+        }
+
+        // At this point we are guaranteed that there is only one leader per slot
+        // because we are operating on committed subdags.
+        assert!(leader_blocks.len() == 1);
+
+        let leader_block = leader_blocks.first().unwrap();
+
+        let decision_blocks = subdag.get_blocks_at_round(decision_round);
+        let mut all_votes = HashMap::new();
+        for potential_cert in decision_blocks {
+            let authority = potential_cert.reference().author;
+            if subdag.is_certificate(&potential_cert, leader_block, &mut all_votes) {
+                tracing::info!(
+                    "Found a certificate {} for leader {leader_block} from authority {authority}",
+                    potential_cert.reference()
+                );
+                tracing::info!(
+                    "[{}] scores +1 reputation for {authority}!",
+                    subdag.context.own_index
+                );
+                scores_per_authority[authority] += 1;
+            }
+        }
+
+        scores_per_authority
+    }
+
+    fn leader_scoring_round_range(&self, min_round: u32, max_round: u32) -> Range<u32> {
+        assert!(min_round < max_round - 1);
+        min_round..max_round.saturating_sub(1)
+    }
+}
```

Comment:
I agree I think we need to conclude on one, but the reason I kept them all was so that we could get the feature in and then test them all thoroughly later and choose a strategy with more information. Do you want me to remove it now or are you okay merging with them all for now.